src/common/UniqueAccess.hpp

```diff
@@ -48,10 +42,13 @@ class AccessGuard
 
 private:
     T *element_{};
-    std::mutex *mutex_{};
-    bool isValid_{true};
+    LockType lock_;
 };
 
+template <typename T>
+using SharedAccessGuard =
+    AccessGuard<const T, std::shared_lock<std::shared_mutex>>;
```

Comment:
I would remove the const here

Reply:
Going to keep the `const` here as per our discussion in chat since without it it's easy to make a difficult to read compile error when working with `SharedAccessGuard`